### PR TITLE
Replace GraphMap’s map with a moka::sync::Cache

### PR DESCRIPTION
**Pull request by chirino**
Fixes issue https://github.com/trustification/trustify/issues/1218

### CODE REVIEW

**Comment by chirino** [comment]
We should add some tests around the cache eviction edge cases:
* Verify old sboms are dropped as the cache fills
* Find out what happens when the cache is smaller a single request requires.

File: modules/analysis/src/model.rs [hunk 1]
```diff
@@ -35,7 +37,35 @@ pub struct PackageNode {
...
+            + self.published.len()
+            + self.document_id.len()
+            + self.product_name.len()
+            + self.product_version.len();
+
+        PackageNode {
+            approximate_memory_size: size.try_into().unwrap_or(u32::MAX),
+            ..self.clone()
+        }
+    }
```
Comment (chirino) [inline modules/analysis/src/model.rs:66]: Is there a better way to estimate how much memory this thing uses up?

**Comment by chirino** [comment]
@JimFuller-RedHat I think I'm happy with this now.  I'm going to keep it draft until #1217 goes in so you guys don't suffer the conflicts.  But if you have idle time maybe worth a gander in case you think this went in the wrong direction.

File: modules/analysis/src/model.rs (hunk 1 above)
Comment (chirino) [reply modules/analysis/src/model.rs:66]: Kinda assuming that saving the approximate_memory_size will give us savings later, it might not.

Comment (chirino) [reply modules/analysis/src/model.rs]: I switched to using the DeepSizeOf train which has a derive macro and it seems to give much better results.

**Comment by JimFuller-RedHat** [comment]
@chirino this looks really great (you are amazing!) ... and yes its likely to change a little bit after our PR (will review in full once we get that one into main) ... one thought was do we need to expose an env var to control params for easy configuration ?

File: modules/analysis/src/config.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    let cfg = parse_size::Config::new().with_binary();
+    cfg.parse_size(s)
+}
+
+#[derive(clap::Args, Debug, Clone)]
+pub struct AnalysisConfig {
+    #[arg(
+        id = "max-cache-size",
+        long,
+        env = "TRUSTD_MAX_CACHE_SIZE",
```
Comment (JimFuller-RedHat) [inline modules/analysis/src/config.rs:12]: super!

File: Cargo.toml [hunk 3]
```diff
@@ -54,12 +54,14 @@ bytes = "1.5"
 bytesize = "1.3"
 chrono = { version = "0.4.35", default-features = false }
 clap = "4"
+moka = "0.12.10"
```
Comment (JimFuller-RedHat) [inline Cargo.toml:57]: have you consider also exposing moka expiration policies TTL and TTI as env vars ? I was hoping that we can control both time and mem dimensions ... Moka looks great!

File: modules/analysis/src/service/load.rs [hunk 4]
```diff
@@ -323,19 +327,25 @@ impl AnalysisService {
         // Set the result. A parallel call might have done the same. We wasted some time, but the
         // state is still correct.
 
-        self.graph.write().insert(distinct_sbom_id.to_string(), g);
+        let g = Arc::new(g);
+        self.graph_cache
```
Comment (JimFuller-RedHat) [inline modules/analysis/src/service/load.rs:331]: it occurred to me there might be times when we want to suspend caching (might be useful for debugging for example)... not sure if this is important or not to consider for right now

File: modules/analysis/src/service/mod.rs [hunk 5]
```diff
@@ -38,7 +38,7 @@ use uuid::Uuid;
 
 #[derive(Clone)]
 pub struct AnalysisService {
-    graph: Arc<RwLock<GraphMap>>,
+    graph_cache: Arc<GraphMap>,
```
Comment (JimFuller-RedHat) [inline modules/analysis/src/service/mod.rs:41]: I would wait till @ctron reviews ... previously he reworked things in this area and it was a bit tricky.

File: Cargo.toml (hunk 3 above)
Comment (chirino) [reply Cargo.toml:57]: Yeah was considering it.   But at least as this PR is, we should be able to avoid blowing up the memory usage which was the main goal of this PR.

File: modules/analysis/src/service/test.rs [hunk 6]
```diff
@@ -354,6 +355,45 @@ async fn test_status_service(ctx: &TrustifyContext) -> Result<(), anyhow::Error>
...
+    let small_sbom_size = service.cache_size_used();
+    assert!(small_sbom_size > 6 * kb);
+    assert!(small_sbom_size < 7 * kb);
+
+    ctx.ingest_documents(["spdx/quarkus-bom-3.2.11.Final-redhat-00001.json"])
+        .await?;
+    let all_graphs = service.load_all_graphs(&ctx.db).await?;
+    assert_eq!(all_graphs.len(), 2);
+
+    let big_sbom_size = service.cache_size_used() - small_sbom_size;
```
Comment (JimFuller-RedHat) [inline modules/analysis/src/service/test.rs:379]: very nice !

**Review by JimFuller-RedHat — APPROVED** [review]
apart from a few minor observations - looks great to me - as @ctron did the final fixes on the service invoke stuff good to get his eyeballs on it as well (as it was more then a little fiddly).

File: modules/analysis/src/config.rs [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+        default_value = "200 MB",
+        help = "Maximum size of the graph cache.  You can use binary units like 'kb', 'mb' or 'gb'.",
+        value_parser = u64_from_size
```
Comment (ctron) [inline modules/analysis/src/config.rs:15]: I'd like to keep this aligned with the existing way to parse byte sizes: https://github.com/trustification/trustify/blob/a7e2855368841cd6e667b35aab5608d411330749/common/infrastructure/src/app/http.rs#L134-L141

File: modules/analysis/src/model.rs [hunk 8]
```diff
@@ -4,8 +4,14 @@ pub use roots::*;
 
 use petgraph::Graph;
 use serde::Serialize;
-use std::ops::{Deref, DerefMut};
-use std::{collections::HashMap, fmt};
+use std::{
+    fmt,
+    ops::{Deref, DerefMut},
+};
+
```
Comment (ctron) [inline modules/analysis/src/model.rs:11]: Please merge and nest this.

File: modules/analysis/src/service/mod.rs [hunk 9]
```diff
@@ -217,31 +237,24 @@ impl AnalysisService {
     fn query_graphs<'a, F>(
         &self,
         query: impl Into<GraphQuery<'a>> + Debug,
-        distinct_sbom_ids: Vec<String>,
+        graphs: Vec<(String, Arc<PackageGraph>)>,
```
Comment (ctron) [inline modules/analysis/src/service/mod.rs:240]: It looks like it's not necessary to pass a `Vec`, so maybe use `&[Arc<PackageGraph>]` instead.

File: modules/analysis/src/service/render.rs [hunk 10]
```diff
@@ -1,7 +1,7 @@
 use super::*;
 
 impl AnalysisService {
-    pub fn render_dot(&self, sbom: &str) -> Option<String> {
+    pub fn render_dot(&self, graph: Arc<PackageGraph>) -> Option<String> {
```
Comment (ctron) [inline modules/analysis/src/service/render.rs:4]: Is the `Arc` really required? Or is a `&PackageGraph` good enough?

File: server/src/profile/api.rs [hunk 11]
```diff
@@ -289,6 +295,7 @@ impl InitData {
         };
 
         Ok(InitData {
+            analysis: AnalysisService::new_sized(run.analysis.max_cache_size),
```
Comment (ctron) [inline server/src/profile/api.rs:298]: I think it would be great it one could construct an `AnalysisService` from an `AnalysisConfig` directly.

File: modules/analysis/src/service/mod.rs [hunk 12]
```diff
@@ -129,20 +129,48 @@ impl AnalysisService {
...
+    /// Default cache size is 100MB.
+    ///
     /// Therefore, it is ok to create a new instance. However, if you want to make use of the
     /// caching, it is necessary to re-use that instance.
     ///
     /// Also, we do not implement default because of this. As a new instance has the implication
     /// of having its own cache. So creating a new instance should be a deliberate choice.
     #[allow(clippy::new_without_default)]
     pub fn new() -> Self {
+        Self::new_sized(1024 * 1024 * 100)
```
Comment (ctron) [inline modules/analysis/src/service/mod.rs:141]: I am worries this gets overlooked. It looks like we use this for tests only. One way could be to add `#[cfg(test]`. Or drop it an have a `const` default value for tests.

**Review by ctron — CHANGES_REQUESTED** [review]
Looks good. A few things to iron out.

And I really would to have more nested `use` sections with every PR rather than less :wink:

File: modules/analysis/src/config.rs (hunk 2 above)
Comment (helio-frota) [reply modules/analysis/src/config.rs:12]: nit mind to add a line on https://github.com/trustification/trustify/blob/main/docs/env-vars.md ?

Comment (chirino) [reply modules/analysis/src/config.rs:9]: will do

File: modules/analysis/src/config.rs (hunk 7 above)
Comment (chirino) [reply modules/analysis/src/config.rs]: will do

File: modules/analysis/src/model.rs (hunk 8 above)
Comment (chirino) [reply modules/analysis/src/model.rs:12]: ok.  still wish tooling automated this.

File: modules/analysis/src/service/mod.rs (hunk 12 above)
Comment (chirino) [reply modules/analysis/src/service/mod.rs]: I'll just drop it and fix it in the tests.

File: modules/analysis/src/service/render.rs (hunk 10 above)
Comment (chirino) [reply modules/analysis/src/service/render.rs]: ref is good enough.

File: modules/analysis/src/service/mod.rs [hunk 13]
```diff
@@ -217,31 +220,24 @@ impl AnalysisService {
...
         query: impl Into<GraphQuery<'a>> + Debug,
-        distinct_sbom_ids: Vec<String>,
+        graphs: &[(String, Arc<PackageGraph>)],
         mut f: F,
     ) where
         F: FnMut(&Graph<PackageNode, Relationship>, NodeIndex, &PackageNode, &mut FixedBitSet),
     {
         let query = query.into();
 
         // RwLock for reading hashmap<graph>
```
Comment (ctron) [inline modules/analysis/src/service/mod.rs:230]: There's not rwlock anymore, is there?

File: server/src/profile/api.rs (hunk 11 above)
Comment (chirino) [reply server/src/profile/api.rs]: Will change that around. And give AnalysisConfig a default.

File: modules/analysis/src/service/mod.rs (hunk 13 above)
Comment (chirino) [reply modules/analysis/src/service/mod.rs]: will drop
